app/agent/runtime/qemu/qemu.service.ts

```diff
@@ -0,0 +1,542 @@
+      timeoutResolve(void 0);
+      await timeoutP;
+    }
+    return r;
+  }
+
+  private async sendQMPMessage(sock: Socket, execute: string, args?: any) {
+    const msg = JSON.stringify({
+      execute,
+      arguments: args,
+    });
+    this.logger.debug(`[${this.instanceId}] Sent QMP message ${msg}`);
+    sock.write(msg);
+  }
+
+  private async qmpConnect(): Promise<Socket> {
+    const sock = new Socket();
+    await new Promise((resolve, reject) => {
+      sock.on("error", (err) => reject(new QmpConnectionFailedError(err.message)));
+      sock.connect(this.qmpSocketPath, () => resolve(void 0));
+    });
+    if ((await this.waitForQMPMessageWithTimeout(sock)) === void 0) {
+      throw new QmpConnectionFailedError("Failed to connect to QMP");
+    }
+    await this.sendQMPMessage(sock, "qmp_capabilities");
+    if ((await this.waitForQMPMessageWithTimeout(sock)) === void 0) {
+      throw new QmpConnectionFailedError("Failed to get qmp capabilities");
+    }
+    return sock;
+  }
+
+  /**
+   * Only works if the VM kernel is compiled with support for
+   * `CONFIG_SERIO_I8042` and `CONFIG_KEYBOARD_ATKBD` as per
+   * https://github.com/firecracker-microvm/firecracker/blob/2b8ad83629af511f918d616aa1c0d441e52c397a/docs/api_requests/actions.md#intel-and-amd-only-sendctrlaltdel
+   *
+   * Waits for the VM to shutdown.
+   */
+  private async shutdownVM(): Promise<void> {
+    let sock: Socket;
+    try {
+      sock = await this.qmpConnect();
+    } catch (err: any) {
+      if (err instanceof QmpConnectionFailedError) return;
+      throw err;
+    }
+    await this.sendQMPMessage(sock, "send-key", {
+      keys: [
+        { type: "qcode", data: "ctrl" },
+        { type: "qcode", data: "alt" },
+        { type: "qcode", data: "delete" },
+      ],
+    });
+    if ((await this.waitForQMPMessageWithTimeout(sock)) === void 0) {
+      throw new Error("Failed to press CTRL+ALT+DELETE");
+    }
+    // Now wait up to 5 seconds for a QMP event that the VM powered down
+    let cleanShutdownDone = false;
+    let t1 = performance.now();
+    while (!cleanShutdownDone && performance.now() - t1 < 5000) {
+      const m = await this.waitForQMPMessageWithTimeout(sock, 100);
+      if (m !== void 0 && (m as any).event === "SHUTDOWN") {
+        cleanShutdownDone = true;
+      }
+    }
+
+    if (!cleanShutdownDone) {
+      this.logger.error("Guest OS refused to shutdown gracefully within 5s. Killing the VM");
+      // In case the VM does not cooperate ask qemu to exit forcefully
+      await this.sendQMPMessage(sock, "quit");
+      // We don't wait for a response cause we may get EOF in the middle of getting the response
+    }
+    sock.destroy();
+
+    const vmInfo = await this.readVmInfoFile();
+    if (vmInfo === null) return;
+    // Give qemu 0.5s to terminate, if qemu failed to stop then send SIGKILL
+    let t2 = performance.now();
+    while (
+      (await doesFileExist(path.join("/proc", vmInfo.pid.toString()))) &&
```

Comment:
Please turn `doesFileExist(path.join("/proc", vmInfo.pid.toString()))` into a closure so it is not repeated 5 times

Reply:
Introduced a `isProcessAlive(pid)` helper